subsys/usb/usbd/class/loopback.c

```diff
@@ -0,0 +1,377 @@
+{
+	switch (event) {
+	case USBD_PME_DETACHED:
+		LOG_DBG("USB detached");
+		break;
+	case USBD_PME_SUSPEND:
+		LOG_DBG("USB suspend");
+		break;
+	case USBD_PME_RESUME:
+		LOG_DBG("USB resume");
+		break;
+	default:
+		break;
+	}
+}
+
+/*
+ * Configuration has been changed, information about it
+ * can be obtained from setup packet.
+ * This can be used to restart transfers if necessary.
+ *
+ * Buffer handling here is a temporary solution
+ * until the driver API was revised.
+ */
+static void lb_cfg_update(struct usbd_class_ctx *cctx,
+			  struct usb_setup_packet *setup)
+{
+	struct net_buf *buf;
+	uint8_t ep = 0;
+	uint16_t mps = 0;
+	struct loopback_desc *lb_desc;
+
+	lb_desc = (struct loopback_desc *)cctx->class_desc;
+
+	switch (setup->bRequest) {
+	case USB_SREQ_SET_CONFIGURATION:
+		LOG_WRN("Configured");
+		if (cctx_restart_out_eps(cctx, 0, true) != 0) {
+			LOG_ERR("Failed to restart transfer");
+		}
+
+		break;
+	case USB_SREQ_SET_INTERFACE:
+		LOG_WRN("Interface %u alternate %u changed",
+			setup->wIndex, setup->wValue);
+		if (cctx_restart_out_eps(cctx, setup->wIndex, false) != 0) {
+			LOG_ERR("Failed to restart transfer");
+		}
+
+		ep = lb_desc->if0_in_ep.bEndpointAddress;
+		mps = lb_desc->if0_in_ep.wMaxPacketSize;
+		buf = usbd_tbuf_alloc(ep, mps);
+		if (buf == NULL) {
+			LOG_ERR("Failed to allocate buffer");
+			break;
+		}
+
+		net_buf_add_mem(buf, lb_buf, mps);
+		usbd_tbuf_submit(buf, false);
+
+		break;
+	}
+}
+
+/*
+ * Common handler for all endpoints.
+ *
+ * Buffer handling here is a temporary solution
+ * until the driver API was revised.
+ */
+static void lb_ep_event_handler(struct usbd_class_ctx *cctx,
+				struct net_buf *buf, int err)
+{
+	struct usbd_buf_ud *ud = NULL;
+	uint8_t ep = 0;
+	uint16_t mps = 0;
+	struct loopback_desc *lb_desc;
+	struct net_buf *new_buf;
+
+	ud = (struct usbd_buf_ud *)net_buf_user_data(buf);
+	lb_desc = (struct loopback_desc *)cctx->class_desc;
+	LOG_DBG("-> ep 0x%02x, len %u, err %d", ud->ep, buf->len, err);
+
+	if (err != 0) {
+		LOG_ERR("Transfer failed with %d", err);
+		net_buf_unref(buf);
+		return;
+	}
+
+	if (ud->ep == lb_desc->if0_out_ep.bEndpointAddress) {
+		ep = lb_desc->if0_out_ep.bEndpointAddress;
+		mps = lb_desc->if0_out_ep.wMaxPacketSize;
+		memcpy(lb_buf, buf->data, buf->len);
+	} else if (ud->ep == lb_desc->if2_out_ep.bEndpointAddress) {
+		ep = lb_desc->if2_out_ep.bEndpointAddress;
+		mps = lb_desc->if2_out_ep.wMaxPacketSize;
+	} else if (ud->ep == lb_desc->if0_in_ep.bEndpointAddress) {
+		ep = lb_desc->if0_in_ep.bEndpointAddress;
+		mps = lb_desc->if0_in_ep.wMaxPacketSize;
+	} else if (ud->ep == lb_desc->if2_in_ep.bEndpointAddress) {
+		ep = lb_desc->if2_in_ep.bEndpointAddress;
+		mps = lb_desc->if2_in_ep.wMaxPacketSize;
+	} else {
+		LOG_ERR("Unknown endpoint, skip");
+		net_buf_unref(buf);
+		return;
+	}
+
+	net_buf_unref(buf);
+
+	new_buf = usbd_tbuf_alloc(ep, mps);
+	if (new_buf == NULL) {
+		LOG_ERR("Failed to allocate buffer");
+		net_buf_unref(buf);
```

Comment:
This is probably a missclick as the `buf` was unrefed just before allocating `new_buf`.